app/Jobs/Api/AddTags.php

```diff
@@ -51,54 +53,20 @@ public function handle ()
 
         $photo = Photo::find($this->request['photo_id']);
 
-        $schema = LitterTags::INSTANCE()->getDecodedJSON();
+        $tags = $this->request['litter'];
 
-        $litterTotal = 0;
-        foreach ($this->request['litter'] as $category => $items)
-        {
-            foreach ($items as $column => $quantity)
-            {
-                // Column on photos table to make a relationship with current category eg smoking_id
-                $id_table = $schema->$category->id_table;
-
-                // Full class path
-                $class = 'App\\Models\\Litter\\Categories\\'.$schema->$category->class;
-
-                // Create reference to category.$id_table on photos if it does not exist
-                if (is_null($photo->$id_table))
-                {
-                    $row = $class::create();
-                    $photo->$id_table = $row->id;
-                    $photo->save();
-                }
-
-                // If it does exist, get it
-                else $row = $class::find($photo->$id_table);
-
-                // Update quantity on the category table
-                $row->$column = $quantity;
-                $row->save();
-
-                // Update Leaderboards if user has public privacy settings
-                // todo - save data per leaderboard
-                if (($user->show_name) || ($user->show_username))
-                {
-                    $country = Country::find($photo->country_id);
-                    $state = State::find($photo->state_id);
-                    $city = City::find($photo->city_id);
-                    Redis::zadd($country->country.':Leaderboard', $user->xp, $user->id);
-                    Redis::zadd($country->country.':'.$state->state.':Leaderboard', $user->xp, $user->id);
-                    Redis::zadd($country->country.':'.$state->state.':'.$city->city.':Leaderboard', $user->xp, $user->id);
-                }
-
-                $litterTotal += $quantity;
-            }
-        }
+        /** @var AddTagsToPhotoAction $addTagsAction */
+        $addTagsAction = app(AddTagsToPhotoAction::class);
+        $litterTotals = $addTagsAction->run($photo, $tags);
 
-        $user->xp += $litterTotal;
+        $user->xp += $litterTotals['all'];
```

Comment:
We do that $user->update() because it doesn't actually retrieve the user's xp from db, but instead lets MySql update it. It is only used on PhotosController and ApiPhotosController because only at that place we have the chance of uploading multiple photos at once, and if we retrieve the xp we get old/inacurate values. On other places we do simple updates because we're sure no concurrency issues might occur.

I've been thinking about extracting that assignXp action, it would be a little odd to have an action to update the user's XP, but on the same place in code update the other fields, like total_litter etc. So I've thought of a more generic action like UpdateUserAction, but that still didn't fit well. I still haven't found a suitable solution for this, but if we're going to migrate the XP on Redis, we can just refactor this when we do the migration, no need to do double work, or solve issues that will not even exist on Redis. What do you think?

Reply:
Something like `User/AssignXp` sounds good to me just to put the action in a shared location. The Redis code will live here in place of updating the user object which handles incrementing well